### DISCUSSION TITLE
Issue with OOM Killer occurring when creating a massive stream

### DISCUSSION THREAD

**Discussion by cold01water** [openobserve/openobserve#1002] [Q&A]
I am currently testing how much log processing my low-spec machine can handle. The machine has 4 cores and 4GB of memory. I am using OpenObserver v0.4.7.

I sent approximately 10,000 messages per second and continued to accumulate logs in a single stream. At around 80GB in actual size and approximately 3.5GB after compression, an OOM Killer event occurred when I attempted a search.

At that time, the WAL size had become huge, and it seemed that the saving process to the stream was not working properly. Even after stopping the log sending, the situation did not change.

Restarting the process improved the situation, but when I performed a search while saving from WAL to the stream, the OOM Killer event occurred again. There have been no issues once the WAL is empty.

Is this the intended behavior? Also, is there a specific limit to the size of a single stream in relation to machine memory?

Note: The number of messages is 2.5 billion.

**Comment by prabhatsharma** [comment]
2 areas are mentioned in your post:

1. Ingestion
2. Search

In order to provide better durability and performance, OpenObserve stores data with minimal processing in WAL. This process is very fast. It allows for quickly ingesting large amounts of data. Much higher than it would have been possible with real time processing and allows for any short bursts of incoming data to be handled well. WAL is stored in JSON files currently (We are working on to change the WAL format to a more efficient format which should be available in subsequent releases). WAL also allows batching of data before pushing on to object storage. 

Batching, conversion to parquet, compression and moving of data to object store is comparatively more compute intensive and happens asynchronously. 

What you experienced was that you pushed OpenObserve a lot more data than it could handle with available hardware resources. OpnObserve tried to handle it with WAL for quite a while but crumbled at some point. Disk speeds can also hamper WAL creation and movement and can be a bottleneck. If performance is your priority then you can enable memory based WAL in OpenObserve. It's a tradeoff of performance vs durability though. 

Search is done by loading data in memory for faster retrieval . We try to use 50% of available memory (configurable by env variable) on the machine for this as of v0.4.7 . 50% memory for search + 50%+ for batching, converting, compressing and moving could have caused OOM. We have made improvements in this area that should be available in the coming release.

> Also, is there a specific limit to the size of a single stream in relation to machine memory?

Not really. You can have petabytes of data in a single stream. What is important is how much data are you processing at any given point in time.

**Comment by cold01water** [comment]
Thanks for the answer.

The movement was generally as expected.
In other words, in the input processing of logs, it is sufficient to receive a quantity of logs that does not exceed 50% of the memory. By doing so, if the search process fits within the memory, it will take up to a maximum of 50%, and if not, it will be read directly from the disk.
Please let me know if my understanding is wrong.

However, I'm curious about what the search speed depends on when a large size is accumulated in a single stream.

When there is a storage of 80GB (compressed: 3.5G), searching from all records takes about 20-30 seconds. Can this be improved only by improving disk IOPS and CPU performance? Or should the stream be divided?

There seems to be a feature called partitioning key, but since there is no explanation yet, I don't fully understand how it works. Judging from the structure of the data directory when partitioning is enabled, I think it is a feature for speeding up searches... like the primary key in ClickHouse?

Naturally, adding search conditions such as time can improve the speed.
However, with the current behavior, when the search screen is opened, it performs a search of all records. The drawback is that it takes a long time until conditional searches become available.

**Comment by prabhatsharma** [comment]
> The movement was generally as expected.
In other words, in the input processing of logs, it is sufficient to receive a quantity of logs that does not exceed 50% of the memory. By doing so, if the search process fits within the memory, it will take up to a maximum of 50%, and if not, it will be read directly from the disk.

This is correct.

Search performance depends on the amount of data being searched on. If you look at the way files are stored physically. You will notice:

```
.
├── file_list
│   └── 2023
│       └── 06
│           ├── 20
│           │   └── 15
│           │       └── 7076956468712439808.json.zst
│           └── 23
│               └── 11
│                   ├── 7077965863164444672.json.zst
│                   └── 7077966518776102912.json.zst
└── files
    ├── default
    │   └── logs
    │       └── default
    │           └── 2023
    │               └── 06
    │                   ├── 20
    │                   │   └── 15
    │                   │       └── 7076956468657913857.parquet
    │                   └── 23
    │                       └── 11
    │                           ├── 7077965863135084544.parquet
    │                           └── 7077966518755131393.parquet
    └── org1
        └── logs
            └── default
                └── 2023
                    └── 06
                        └── 20
                            └── 15
                                └── 7076956468657913856.parquet

```

`files` is the location where actual data is stored. You will notice that it is divided into folders based on year/month/day/hour . These are the base partitions. 

Your search performance depends on how much data you are scanning through a given query. If you add a filter to search only for a specific day as opposed to the whole year then your search performance will be better since you are scanning less data. 

In organizations where large amount of data is flowing in, you can create additional partitions -e.g. host_name . This will allow you to have conditions based on the particular host_name e.g. host_name=host1 . This will reduce the amount of data being scanned and will improve search performance. 

You can have multiple partitions. Make sure that when you are creating partitions don't let each data file become too small. Also if you are using s3 then you would want to not have to small files as you pay for each read and write and reading 1 kb file and 100 MB file costs same.  As a general rule of thumb try to have each file between 5-15 MB.

**Comment by cold01water** [comment]
Thank you. I understood about the inner workings.

Sorry for all the questions, but one last thing.
When searching by date, a specific partition is scanned. Is the memory consumption at this time the original size of the log? Is it the size of the file? I guess it depends on the query, but I would like to know how memory is used in a case like a simple select * from db issued.

**Comment by prabhatsharma** [comment]
Data is stored in memory in compressed bytes (Size of actual files). e.g.  You ingested 1 TB of logs. It is 30 GB compressed. The data stored in RAM will be 30 GB (Depends on the amount of RAM you have. We use 50% of RAM as cache...).

When you run a query like `select * from table` then some additional memory will be needed on top of the existing RAM used.
